src/main/webapp/app/exercises/shared/rating/star-rating/star-rating.component.ts

```diff
@@ -0,0 +1,300 @@
+            starElement.title = starElement.dataset.index;
+            starContainer.appendChild(starElement);
+            this.stars.push(starElement);
+        });
+    }
+
+    private applySizeAllStars() {
+        if (this._size) {
+            if (this.stars.length === 0) {
+                this.setStars();
+            }
+            this.stars.forEach((star: HTMLElement) => {
+                const newSize = this.size.match(/\d+/)![0];
+                const halfSize = (parseInt(newSize, 10) * 10) / 24;
+                const halfMargin = 0 - (parseInt(newSize, 10) * 20) / 24;
+                star.style.setProperty(StarRatingComponent.VAR_SIZE, this.size);
+                if (star.classList.contains(StarRatingComponent.CLS_HALF_STAR)) {
+                    star.style.setProperty(StarRatingComponent.VAR_HALF_WIDTH, `${halfSize}px`);
+                    star.style.setProperty(StarRatingComponent.VAR_HALF_MARGIN, `${halfMargin}px`);
+                }
+            });
+        }
+    }
+
+    private applyColorStyleAllStars(setChecked: boolean) {
+        if (this.stars.length === 0) {
+            this.setStars();
+        }
+        this.stars.forEach((star: HTMLElement) => {
+            if (setChecked) {
+                this.applyCheckedColorStyle(star);
+            } else {
+                this.applyUnCheckedColorStyle(star);
+            }
+        });
+    }
+
+    private applyColorStyle(starElement: HTMLSpanElement) {
+        this.applyCheckedColorStyle(starElement);
+        this.applyUnCheckedColorStyle(starElement);
+    }
+
+    private applyCheckedColorStyle(starElement: HTMLSpanElement) {
+        starElement.style.setProperty(StarRatingComponent.VAR_CHECKED_COLOR, this._checkedColor);
+    }
+
+    private applyUnCheckedColorStyle(starElement: HTMLSpanElement) {
+        starElement.style.setProperty(StarRatingComponent.VAR_UNCHECKED_COLOR, this._unCheckedColor);
+    }
+
+    private generateRating(forceGenerate = false) {
+        if (!this.mainElement) {
+            return;
+        }
+        if (this.readOnly && !forceGenerate) {
+            return;
+        }
+
+        if (this.stars.length === 0) {
+            this.setStars();
+        }
+        this.mainElement.nativeElement.title = this.value;
+
+        let hasDecimals: boolean = !!(Number.parseFloat(this.value.toString()) % 1).toString().substring(3, 2);
+
+        let i = 1;
+        this.stars.forEach((star: HTMLElement) => {
+            star.className = '';
+            this.applyColorStyle(star);
+            StarRatingComponent.addDefaultClass(star);
+
+            if (this.value >= i) {
+                // star on
+                StarRatingComponent.addCheckedStarClass(star);
+            } else {
+                // half star
+                if (hasDecimals) {
+                    StarRatingComponent.addHalfStarClass(star);
+                    hasDecimals = false;
+                }
+            }
+            i++;
```

Comment:
```suggestion
        this.stars.forEach((star: HTMLElement, i: number) => {
            star.className = '';
            this.applyColorStyle(star);
            StarRatingComponent.addDefaultClass(star);

            if (this.value > i) {
                // star on
                StarRatingComponent.addCheckedStarClass(star);
            } else {
                // half star
                if (hasDecimals) {
                    StarRatingComponent.addHalfStarClass(star);
                    hasDecimals = false;
                }
            }
```

There should be an overload on `forEach` to also be able to use the index directly.